src/pylibsshext/scp.pyx

```diff
@@ -0,0 +1,132 @@
+                    raise LibsshSCPException("Can't open remote file: %s" % self._get_ssh_error_str())
+
+                # Write to the open file
+                rc = libssh.ssh_scp_write(scp, PyBytes_AS_STRING(f.read()), file_size);
+                if rc != libssh.SSH_OK:
+                    raise LibsshSCPException("Can't write to remote file: %s" % self._get_ssh_error_str())
+            finally:
+                libssh.ssh_scp_close(scp)
+                libssh.ssh_scp_free(scp)
+
+            return libssh.SSH_OK
+
+    def get(self, remote_file, local_file):
+        cdef char *read_buffer
+
+        remote_file_b = remote_file
+        if isinstance(remote_file_b, unicode):
+            remote_file_b = remote_file.encode("utf-8")
+            remote_dir_b, filename_b = os.path.split(remote_file_b)
+
+        # Create the SCP session in read mode
+        scp = libssh.ssh_scp_new(self._libssh_session, libssh.SSH_SCP_READ, remote_file_b)
+        if scp is NULL:
+            raise LibsshSCPException("Allocating SCP session of remote file [%s] for write failed with error [%s]" % (remote_file, self._get_ssh_error_str()))
+
+        # Initialize the SCP channel
+        rc = libssh.ssh_scp_init(scp)
+        if rc != libssh.SSH_OK:
+            libssh.ssh_scp_free(scp)
+            raise LibsshSCPException("Initializing SCP session of remote file [%s] for write failed with error [%s]" % (remote_file, self._get_ssh_error_str()))
+
+        try:
+            # Request to pull the file
+            rc = libssh.ssh_scp_pull_request(scp)
+            if rc != libssh.SSH_SCP_REQUEST_NEWFILE:
+                raise LibsshSCPException("Error receiving information about file: %s" % self._get_ssh_error_str())
+
+            size = libssh.ssh_scp_request_get_size(scp)
+            mode = libssh.ssh_scp_request_get_permissions(scp)
+
+            read_buffer = <char *>PyMem_Malloc(size)
+            if read_buffer is NULL:
+                raise LibsshSCPException("Memory allocation error")
+
+            # Indicate the transfer is ready to begin
+            libssh.ssh_scp_accept_request(scp)
+            # Read the file
+            rc = libssh.ssh_scp_read(scp, read_buffer, size)
+            if rc == libssh.SSH_ERROR:
+                PyMem_Free(read_buffer)
+                raise LibsshSCPException("Error receiving file data: %s" % self._get_ssh_error_str())
+
+            with open(local_file, "w+") as f:
+                f.write(read_buffer.decode("utf-8"))
+            PyMem_Free(read_buffer)
+
+            # Make sure we have finished requesting files
+            rc = libssh.ssh_scp_pull_request(scp)
+            if rc != libssh.SSH_SCP_REQUEST_EOF:
+                raise LibsshSCPException("Unexpected request: %s" % self._get_ssh_error_str())
+
+        finally:
+            libssh.ssh_scp_close(scp)
+            libssh.ssh_scp_free(scp)
+
+        return libssh.SSH_OK
+
+    def close(self):
+        if self._libssh_session is not NULL:
```

Comment:
I think this `close` method should handle freeing of `scp` resources that are still allocated.

Reply:
Hm, if `_libssh_session` is handled by `session`'s `__dealloc__`, I don't think there is anything left to free as the scp sessions seem to be single-purpose and get freed at the end of a call.

Is there anything I'm missing that needs to be here, or just drop it like `__dealloc__`?

Reply:
Yeah if it not required we can drop it